Event detail for manually normalized files can be manually specified

Hwesta opened the pull request:
Manually normalized files can have the event detail populated by a person.  Form becomes available for manually normalized files after their are discovered (after normalization in most cases, after 'approve normalization' if 'manually normalize' was selected).  New waiting point for this information to be entered added.

Will need to be updated with preconfigured choice changes from #2.

Hwesta committed:
Update store AIP for Storage Service changes

Storage service requires pipeline URI for creating a package.  Also added
error printing to the storeAIP script.

Note that this still works if the storage service has not been update.

Hwesta committed:
Add user-specified manual normalization event details

refs #5216

Add a form to enter event detail information for manually normalized files
after Normalization takes place.  Only works for files that were normalized
prior to Transfer.

Update normalize.py to add Derivation and Event for manually normalized
files.  Files are left in the manualNormalization directory so that redo
normalization doesn't delete them with Archivematica-generated files.
Update the cleanup script to remove duplicate Events and Derivations. Update
scipt that associates original and manual normalized files to modify an
existing Event, if it can find one.

Add Events & Derivations to models.py.

Hwesta committed:
Add wait point for entering manual normalization metadata

refs #5216

Add a 'user choice' after handling manually normalized files, so the system
pauses until manual normalization event detail information has been
entered. Since this is not a real choice, use set replacement dict with no
real value in replacement dict.

Hwesta committed:
Add preconfigured choice to skip entering manual normalization metadata

refs #5216

Add 'skip waiting to allow manual normalization of metadata entry' to
preconfigured choices, and set the default to yes (skip waiting).

Commit:
Add metadata entry reminder at the end of transfer

This adds an optional new processing step that prompts users to add
Dublin Core metadata at the last possible step before the AIP METS file
is written. It can be disabled via the processing configuration screen.

Refs #6217

mistydemeo commented:
Careful committing empty `.gitignores`. ;)

mistydemeo commented on src/MCPClient/lib/clientScripts/storeAIP.py:
```diff
@@ -73,7 +73,7 @@ def store_aip(aip_destination_uri, aip_path, sip_uuid, sip_name):
         size=os.path.getsize(aip_path)
     )
-    if new_file is not None and new_file['status'] != "FAIL":
+    if new_file is not None and new_file.get('status', '') != "FAIL":
```
Is this meant to be true if `new_file` doesn't have a `'status'` attribute?

Hwesta replied:
Yes, for compatibility if the storage service hasn't been updated to return a 'status'.  Possibly not an issue now, but it was when I wrote it.

mistydemeo commented on src/MCPClient/lib/clientScripts/normalize.py:
```diff
@@ -117,7 +117,8 @@ def check_manual_normalization(opts):
             print('Looking for', filename, 'in database')
-            return File.objects.get(sip=opts.sip_uuid, originallocation__iendswith=filename).currentlocation #removedtime = 0
+            # FIXME: SQL uses removedtime=0. Convince Django to express this
+            return File.objects.get(sip=opts.sip_uuid, originallocation__iendswith=filename) #removedtime = 0
```
Maybe tidy up the comments since you're touching that part of the code?

Hwesta replied:
I added the FIXME and #removedtime = 0 comments, since I couldn't find a way to filter DateTimeFields on '0', which is what they're set to.  It's intended to find files that have not yet been removed.  Suggestions for a fix welcome!

mistydemeo commented on src/MCPClient/lib/clientScripts/normalize.py:
```diff
@@ -203,17 +195,43 @@ def once_normalized(command, opts, replacement_dict):
+    """ Add the derivation link for preservation files and the event. """
+    if today is None:
+        today = str(datetime.date.today())
+    #Add event information to current file
```
Ultra nitpick, but space after hash?

mistydemeo commented on src/MCPClient/lib/clientScripts/retryNormalizeRemoveNormalized.py:
```diff
@@ -70,6 +71,26 @@ def removePreservationFiles(SIPDirectory, SIPUUID):
+        databaseInterface.runSQL(sql)
+    except Exception as inst:
+        traceback.print_exc(file=sys.stdout)
+        print type(inst)     # the exception instance
```
Is this necessary if we already printed the traceback? `traceback.print_exc()` should already print the exception class and its message.

Hwesta replied:
Probably code I copied/moved from somewhere else that I didn't clean up - I'll fix it.  (logging.exception is probably more appropriate anyway)

mistydemeo commented on src/dashboard/src/components/ingest/views.py:
```diff
@@ -167,6 +177,27 @@ def ingest_metadata_edit(request, uuid, id=None):
     return render(request, 'ingest/metadata_edit.html', locals())
 
+def ingest_metadata_event_detail(request, uuid):
+    EventDetailFormset = modelformset_factory(models.Event, form=forms.EventDetailForm, extra=0)
```
Break up the line lengths, maybe? Not a big deal though.

mistydemeo commented on src/dashboard/src/main/models.py:
```diff
@@ -132,6 +132,44 @@ class MetadataAppliesToType(models.Model):
+class Event(models.Model):
+    id = models.AutoField(primary_key=True, db_column='pk', editable=False)
+    event_id = UUIDField(auto=False, null=True, db_column='eventIdentifierUUID')
+    file_uuid = models.ForeignKey('File', db_column='fileUUID', to_field='uuid', null=True, blank=True)
```
Underscores or no underscores? IIRC we recently dumped the underscores from a different model.

Hwesta replied:
I'm having convention conflicts where PEP8 says to use underscores (which I prefer), but the existing code doesn't and I can't decide which one overrides in a specific case.

mistydemeo replied:
Underscores always IMO :snake: Maybe we should plan to just gradually introduce more and more underscores and overwhelm the original convention.

mistydemeo commented:
Only minor nitpicks, really. Looks good!

Hwesta committed:
Minor fixes from code review

Hwesta committed:
Update pre-METS creation metadata reminder

refs #6217, #5216

Both 5216 and 6217 require a wait before the AIP METS is created to
allow the user to enter metadata.  Removed the 5216 manually normalized
files only wait, and updated the more general wait to work in all cases.

Moved the wait to before the MicroServiceChainLink split and added a
watched directory, removed the manually normalized files only wait and
associated watched directory.

Hwesta committed:
Move watchedDirectory to correct location

Hwesta commented:
Oops, that .gitignore watchedDirectory was in the wrong place, moved it to the correct one.

A bunch of the code github says is new is actually moved from elsewhere (not nearby) that I didn't clean up.

Hwesta committed:
Minor fixes

Fixed template typo, added display for when formset is empty.

Hwesta committed:
Clean up exception logs

Clean up output when an exception happens, print more detailed information
about error.
